packager/packager.cc

```diff
@@ -67,7 +67,7 @@ class PackageDB final {
 private:
     // The only thread that is allowed write access to this class.
     const std::thread::id owner;
-    vector<shared_ptr<const PackageInfo>> packages;
+    UnorderedMap<std::string, shared_ptr<const PackageInfo>> packageInfoByPathPrefix;
```

Comment:
`GlobalState` contains a map from string to file. Can we use that instead?

https://github.com/sorbet/sorbet/blob/e08172cf/core/GlobalState.h#L151

* Make this a map from `FileRef` to `shared_ptr<const PackageInfo>`.
* Run `gs.findFileByPath` on `$prefix/__packager.rb`
* If the return value `.exists()`, look up the corresponding `PackageInfo` in the `packageInfoByFile` map.

That will avoid redundantly storing these strings another time and will avoid needing this map.

Also, since the string is continually getting shorter during the lookup loop _and_ since `std::string` is mutable, you can perform lookup with only one string allocation:

 * Allocate the string for the file's folder + `__packager`.
 * Perform lookup with `findFileByPath`.
 * If nothing found, then call [resize](https://www.cplusplus.com/reference/string/string/resize/) up to the next `/` and then [append](https://www.cplusplus.com/reference/string/string/append/) with the `__packager.rb` constant. That should avoid re-allocating the string and will write to the same memory buffer.
   * Or, better yet, [replace](https://www.cplusplus.com/reference/string/string/replace/) at the next `/` and _then_ [resize](https://www.cplusplus.com/reference/string/string/resize/) to shrink the string to the last character of `__packager.rb`. In many cases I bet resize won't even be needed.

Reply:
@jvilk-stripe I can look into the `resize` suggestion. With regards to using `findFileByPath` -- the purpose of making this change is to allow us to add more custom paths to a package, some of which won't have a `__package.rb` in them. So I'm not sure if that will work for us.

Reply:
As per out-of-band convos, going to stick with `substr` as `path` is a `string_view` object.